[spec] Reverse subtyping (#110)
dfinity/candid: spec/Candid.md

In all these cases, a receiver who cannot handle the tag or field will simply see `null`.
This allows for maximal flexibility when evolving an interface over time, while still remaining sound.

### Examples

For example, a representative case is an interface of the following form:
```
// Version 1
type t = {x : nat};
service : {
  produce : () -> t;
  consume : t -> ();
}
```
The subtyping rules allow extending type `t` with additional fields later, as long as they are given optional type:
```
// Version 2
type t = {x : nat; y : opt nat};
service : {
  produce : () -> t;
  consume : t -> ();
}
```
Under normal subtyping rules, this wouldn't be allowed, because such record extension isn't usually compatible (sound) when `t` occurs in inbound position, such as with the `consume` function.
There might be existing clients that are not aware of the new fields, and would fail to provide them.
However, by restricting such fields to option types, and interpreting them as `null` when missing, such a mismatch is bridged.

Such extensibility also extends to *higher-order* examples, where functions themselves become parameters:
```
type t = {x : nat};
service : {

>   h1 : (f1 : () -> t) -> ();      // might call f() and expects a t
>   h2 : (f2 : t -> ()) -> ();      // might call f({x = 5})

Reviewer: ```suggestion
  h1 : (f1 : () -> t) -> ();      // might call f1() and expects a t
  h2 : (f2 : t -> ()) -> ();      // might call f2({x = 5})
```

* Higher-order Extensibility: Subtyping extends to higher-order cases, where functions become parameters, so that there is no unique owner for their input/output contract.

* Language Injectivity: Subtyping does not depend on version or role annotations in interfaces that have no counterpart in source languages, i.e., an interface description can be generated from source-level types without special features or cross-version knowledge.

* Simple Deserialisation: Deserialisation does not require (sub)type checks other than the type-directed traversal of the value blob.

* Type Erasure: Deserialised values do not require carrying dynamic type information on the language side. In particular, service and function references can be represented without type information.

* No covert channels: Serialisation never includes any fields in the value that the sender is not aware of. Specifically, when passing on a value to a third party that the sender previously received itself, then that will only contain fields that the sender intends to send out per its type.

> However, something has to give, so one seemingly desirable property that is not maintained is *transitive coherence*, i.e., when deserialising and reserialising a value and then deserialising that again at a different supertype, it might yield another result than it would if the original value was deserialised at the later supertype.

Reviewer 2: (minor tweak)
```suggestion
However, something has to give, so one seemingly desirable property that is not maintained is *transitive coherence*, i.e., given a value serialized at some type, deserialized and serialized at a supertype, and then again deserialized at a supertype of the supertype may yield a diferent result than deserialised directly at the later supertype.
```

Finally, in order to maintain *transitivity* of subtyping, an unusual rule allows, in fact, *any* option type to be regarded as a subtype of any other.
```
not (<datatype> <: <datatype'>)
---------------------------------
opt <datatype> <: opt <datatype'>
```
*Note:* This rule is necessary in the presence of the unusual record and variant rules shown below. Without it, certain upgrades may generally be valid one step at a time, but not taken together, which could cause problems for clients catching up with multiple upgrades.
For example, given a record type `record {666 : opt nat}` it is valid to remove the field `666` by the rule below and evolve the type to `record {}`.
A later step might legally re-add a field of the same name but with a different type, producing, e.g.,`record {666 : opt text}`.
A client having missed the intermediate step will have to upgrade directly from the original to the newest version of the type.
If two option type do not match up, its value will be treated as `null`.

> In practice, users are strongly discouraged to ever remove a record field or ariant tag and later re-add it with a different meaning.

Reviewer: ```suggestion
In practice, users are strongly discouraged to ever remove a record field or a variant tag and later re-add it with a different meaning.
```

Finally, in order to maintain *transitivity* of subtyping, an unusual rule allows, in fact, *any* option type to be regarded as a subtype of any other.
```
not (<datatype> <: <datatype'>)
---------------------------------
opt <datatype> <: opt <datatype'>
```
*Note:* This rule is necessary in the presence of the unusual record and variant rules shown below. Without it, certain upgrades may generally be valid one step at a time, but not taken together, which could cause problems for clients catching up with multiple upgrades.
For example, given a record type `record {666 : opt nat}` it is valid to remove the field `666` by the rule below and evolve the type to `record {}`.
A later step might legally re-add a field of the same name but with a different type, producing, e.g.,`record {666 : opt text}`.
A client having missed the intermediate step will have to upgrade directly from the original to the newest version of the type.
If two option type do not match up, its value will be treated as `null`.

> In practice, users are strongly discouraged to ever remove a record field or a variant tag and later re-add it with a different meaning.

Reviewer 2: Wouldn’t this work:
```suggestion
In practice, users are strongly discouraged to ever remove an optional record field or a variant tag and later re-add it with a different meaning. Instead of removing an optional record field, it should be replaced with `opt empty`, to prevent re-usd of that field.
```
Maybe we can introduce an alias for `opt empty` to make this less obscure?

This is also how I envisioned the restricted “sensible evolution” relation to look like: It would not allow removing optional fields, only “disabling” them in this way.

This might require a form of polarities in the rules for “can evolve to”. So why do I think this might work (when it didn’t work for normal subtyping)? Because in the higher order case, where two services evolve independently, the composition needs to work with regard to the unrestricted subtyping relation, which is larger.

Author: Hm, I'm not convinced that wouldn't run afoul all the same problems in the higher-order or third-party cases. But happy to explore, later. :)

record { <nat> : <datatype>; <fieldtype>;* } <: record { <nat> : <datatype'>; <fieldtype'>;* }
```

In order to be able to evolve and extend record types that also occur in inbound position (i.e., are used both as function results and function parameters), the subtype relation also supports *removing* fields from records, provided they are optional.
```
record { <fieldtype>;* } <: record { <fieldtype'>;* }
----------------------------------------------------------------------------------------------
record { <fieldtype>;* } <: record { <nat> : opt <datatype'>; <fieldtype'>;* }
```

> *Note:* This rule is unusual form a regular subtyping perspective, but necessary in practice.

Reviewer: ```suggestion
*Note:* This rule is unusual from a regular subtyping perspective, but necessary in practice.
```

variant { <nat> : <datatype>; <fieldtype>;* } <: variant { <nat> : <datatype'>; <fieldtype'>;* }
```

In order to be able to evolve and extend variant types that also occur in outbound position (i.e., are used both as function results and function parameters), the subtype relation also supports *adding* tags to variants, provided the variant itself is optional.
```
opt variant { <fieldtype>;* } <: opt variant { <fieldtype'>;* }
----------------------------------------------------------------------------------------------
opt variant { <nat> : opt <datatype>; <fieldtype>;* } <: opt variant { <fieldtype'>;* }

> ```

Author: > Ah, but they have different elaboration of course… this makes your rules overlapping (I think you avoided that before…) A problem?

Yes, it would be better if this wasn't there. Same problem as above. For now, I added similar handwaving to pepper over the non-determinism.

> But I thought we’d agree we need different relations for “sensible evolution” and “decoding”?

I realised that doesn't actually work. You need to allow all of decoding in the evolution check, because cases like that might not be under your own control: some type you use may originate from another canister, that has performed multiple evolutionary steps that are transitively non-coherent, but you have missed the intermediate step that would avoid the "ugly" transitive case when you upgrade yourself.

Hence the informal language in L785+ instead, about discouraging users to create that case and implementations to warn about it.

IOW, we need the completeness property after all, even if some cases hopefully should never occur in practice if everybody follows good style.

> ```

Reviewer 2: I don't follow the example. Even if you miss a step, and even if there there is incoherence, the “interface evolution” relation would still be transitive, woudn’t it?  But maybe there is some complicate higher-order case where some canister is in the unfortunate situation where it is force to upgrade it’s interface in an undesirable way.

> ```

Author: Our previous conclusion was (unless I misremember something) that the transitive rule (opt t <: opt t' with incompatible t, t') should not be allowed in the upgrade check, because that check always deals with a single evolutionary step where you simply shouldn't be doing that. And that we hence should define two relations, one with and the other without this rule (or something like that).

But the assumption that the check always deals with a single evolutionary step is bogus if your interface has to adapt to (the latest version of) somebody else's interface types, whose evolution might have a different pace, and who makes bad transitive evolutionary choices outside your control. In that case you may need the "bad" rule on options even during upgrade checks.

> ```

Reviewer 2: > Our previous conclusion was (unless I misremember something) that the transitive rule (opt t <: opt t' with incompatible t, t') should not be allowed in the upgrade check, 

yes… but

> because that check always deals with a single evolutionary step where you simply shouldn't be doing that

wasn’t the reason, I think. I thought the reason was that we only had problems when you need the bad rules due to transitivity because you composed two serivces, and _each_ was doing one (or more) sensible steps. But the transtive relation was only relevant for decoding, not an actual service evolution.

That’s why I hope we can have a  “can sensibly evolve” relation (probably with polarities), that’s transitive, and a separate relation “can be decoded at” relation that has the unwanted relations.

But maybe I am too optimistic, and there will always be corner cases where a service needs to change its interface is a “bad” way.

> *Note:* This rule is unusual form a regular subtyping perspective, but it is the dual to the one for records.

Reviewer: ```suggestion
*Note:* This rule is unusual from a regular subtyping perspective, but it is the dual to the one for records.
```

<datatype> <: <datatype'> ~> f
------------------------------------------
<datatype> <: opt <datatype'> ~> \x.?(f x)

not (<datatype> <: <datatype'>)
---------------------------------
opt <datatype> <: opt <datatype'>

>   ~> \x.join_opt (\y.

Reviewer 2: `join_opt` is a bit obscure. maybe write it out with `match … with`, for greater clarity?

Author: Done.

>   ~> \x.join_opt (\y.

Reviewer 2: Yes, we can clean up separtely.

I don’t think it’d be a four-place relation: Didn’t we determine that the input data should be considered untyped (and the type description just be an encoding help)? So I’d expect `v <: v' : t` or maybe more suggestively `v ~> v':t`.

>   ~> \x.join_opt (\y.

Author: I suppose we could phrase it like that, but that would be a bit of a disconnect with the actual serialisation format, which _does_ include a type -- it just may not be "principal".

>   ~> \x.join_opt (\y.

Reviewer 2: But that disconnect would be intentional: Didn't we just last week realize that it was misguided to take that type as more than just a way to help understand the _binary_ structure of the value?

       if (exists <datatype''>.
          y : <datatype''> /\

>           <datatype''> <: <datatype> /\

Reviewer: ```suggestion
          <datatype''> <: <datatype'> /\
```

          <datatype''> <: <datatype> ~> f)
       then ?(f y)
       else null)
```
where
```
v : <datatype>  iff  M(v : <datatype>) is defined
```

> The last rule *optimistically* tries to decode an option value when the types are not statically known to match.

Reviewer 2: There is no “when not statically known to match” due to “Simple Deserialisation: Deserialisation does not require (sub)type checks other than the type-directed traversal of the value blob”

It succceeds if there would have been a valid type for the value (written `v : t`) that is a subtype of both types.
The effect of this rule is that decoders do not need to perform a subtype check during deserialisation.
Instead, they can simply try to deserialise, and if they encounter a mismatch abort, to the innermost option type, returning `null`.

> *Note:* The working assumption is that the type describing its incoming value is typically *principal*, i.e., the most precise type assignable to the value (in a sense that could be made precise).

Reviewer 2: I don’t see right now where we make that assumption, or how it comes into play, but it’s early in the morning… Isn’t the purpose of the whole “if exists” to deal with non-most-precise types?

Anyways, the assumption isn’t even possible (not just costly to enforce): Consider:
```
(vec { opt variant { a } ; opt  variant { a }  }) : (vec ?(variant { a; b }))
```
Clearly the whole message has principal type, but when we reach the optional value `v = opt variant { a }` and want to apply this rule, the type of that value is `?(variant { a; b })` which is not principal, and there is no way to have it principal.

So maybe this Note can be removed? Or maybe I don’t see the point yet.

Author: Wording was probably misleading, tweaked. The "principality" property only needs to hold for the whole type, not inductively for its parts. In your example (I suppose the second value ought to be `b`), `?(variant { a; b })` is the principal type -- and it's enough, since it ensures that both `a` and `b` will ultimately be checked against the respective supertype.

Reviewer 2: Hmm, I think I am beginning to understand why you want this principality here.

This might all be clearer if the rules for decoding would just not look at the input type at all, and only at the values.

> In that case, `<datatype''>` will equal `<datatype>` and deserialisation succeeds exactly if the types match.

Reviewer: ```suggestion
In that case, `<datatype''>` equals `<datatype>` and deserialization succeeds exactly if the types match.
```

------------------------------------------------------------------------------------------------
variant { <nat> : <datatype>; <fieldtype>;* } <: variant { <nat> : <datatype'>; <fieldtype'>;* }
  ~> \x.case x of <nat> y => <nat> (f1 y) | _ => f2 x

opt variant { <fieldtype>;* } <: opt variant { <fieldtype'>;* } ~> f
----------------------------------------------------------------------------------------------
opt variant { <nat> : opt <datatype>; <fieldtype>;* } <: opt variant { <fieldtype'>;* }

>   ~> \x.join_opt (map_opt (\y.case y of <nat> z => null | _ => ?(f x)))

Reviewer 2: This overlaps with the more general `opt` rule. It’s clear that it morally should have precendece, but still, is that a problem?
